sdk/cosmos/azure-cosmos-spark_3-1_2-12/docs/scenarios/Ingestion.md

```diff
@@ -0,0 +1,266 @@
+# Ingesting data into Azure Cosmos DB via the Spark Connector for Spark 3.*
+
+## Introduction
+The `cosmos.oltp` data source can be used to write a data frame into Cosmos DB. By default the ingestion will happen using bulk mode and parameters like the `micro batch size` are automatically modified behind the scenes depending on the rate of throttled requests and transient errors like request timeouts. The `cosmos.oltp` data source will attempt to saturate the available throughput. So, a certain amount of 429 (Throttled requests) are expected and are actually a good sign, because it means, the available throughput is really  saturated. If it is not desired to saturate the entire available throughput, client-side throughput control can be enabled - which means the ingestion will only attempt to saturate the throughput budget provided via throughput control.
+
+### How bulk ingestion works
+In Spark the data frame containing the input data will usually be partitioned. For each partition of the input data Spark will create a Spark task, which will process this partition (a slice of the input data) independently of other Spark tasks.
+
+![Spark partitioning overview](./media/Spark_partitioning.png)
+
+The `cosmos.oltp` data source will create one singleton `CosmosAsyncClient` per Executor/Driver-node - so, this client instance will be shared across multiple Spark tasks being executed concurrently on a single Executor node. But it will create one `BulkWriter` for each Spark task/partition - this `BulkWriter` will only be responsible for ingesting the data of a single Spark partition.
+Each `BulkWriter` will buffer data in-memory for each physical partition in Cosmos DB and "flush" it by sending a single request ingesting multiple documents. The buffers are flushed when certain thresholds are met. Thresholds resulting in "flushing" documents for a physical partition into the Cosmos DB backend are based on the currently determined micro batch size, the total payload size for all buffered documents and an interval, measuring how long the documents buffered have not been persisted yet. This "flush interval" makes sure that even on partitions with just a few documents, the documents get persisted after a couple of seconds at least.
```

Comment:
Is it after a couple of seconds or up to 1 second?

Reply:
Left this intentionally vague because I don't want changes to the flush interval (not exposed in config - so implementation detail) to become breaking changes. The implementation currently uses 1 second as a flush interval up to the point where the Flux with the incoming data signals that there is no more data - then we switch to 100ms. Both might need to be changed - and if necessary we might even need to expose them in config. For Spark it is perfectly fine - but for customers using the Java SDK directly there could be widely different use cases where latency might be critical.

Reply:
Leaving open only to make it easier for reviewers to see the discussion.